modules/gdscript/gdscript_parser.cpp

```diff
@@ -3857,6 +3860,33 @@ bool GDScriptParser::validate_annotation_arguments(AnnotationNode *p_annotation)
 	return true;
 }
 
+bool GDScriptParser::uid_annotation(const AnnotationNode *p_annotation, Node *p_target, ClassNode *p_class) {
+	ERR_FAIL_COND_V_MSG(p_target->type != Node::CLASS, false, R"("@uid" annotation can only be applied to classes.)");
+	ERR_FAIL_COND_V(p_annotation->resolved_arguments.is_empty(), false);
+
+#ifdef DEBUG_ENABLED
+	if (this->_has_uid) {
+		push_error(R"("@uid" annotation can only be used once.)", p_annotation);
+		return false;
+	}
+#endif // DEBUG_ENABLED
+
+	const String &uid_string = p_annotation->resolved_arguments[0];
+#ifdef DEBUG_ENABLED
+	if (ResourceUID::get_singleton()->text_to_id(uid_string) == ResourceUID::INVALID_ID) {
+		push_error(R"(The annotated UID is invalid.)", p_annotation);
+		return false;
+	}
+#endif // DEBUG_ENABLED
+
+	ClassNode *class_node = static_cast<ClassNode *>(p_target);
+	class_node->uid_string = uid_string;
+	class_node->uid_lines = Vector2i(p_annotation->start_line - 1, p_annotation->end_line - 1); // Lines start from 1, so need to subtract.
```

Comment:
It makes sense to move the `uid_lines` assignment above the UID validity check so that the first invalid annotation will be fixed rather than another one added.

https://github.com/godotengine/godot/assets/47700418/56c44f56-bf47-4f89-9e9d-ba213d64ba72

Reply:
Right, but the annotation can be broken in many ways, like by missing `)` or `"`. Is it possible to still reliably get its lines in every case?

Reply:
At this point, the annotation has been successfully parsed and has exactly one argument, which is a string literal. If the syntax is invalid, the annotation has the wrong argument count, or the argument is not a string literal, then the `uid_annotation()` method will not be called.